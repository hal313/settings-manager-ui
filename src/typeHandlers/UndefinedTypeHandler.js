import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';

export class UndefinedTypeHandler {

    constructor() {}

    getType() {
        return 'undefined';
    }

    getValue(/*element*/) {
        return undefined;
    }

    setValue(element/*, value*/) {
        element.value = undefined;
    }

    createElement(name/* , value */) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('div');
        element.setAttribute('name', name);
        element.setAttribute('value', undefined);
        element.setAttribute(Constants.ATTRIBUTE_NAME, name);
        element.setAttribute(Constants.ATTRIBUTE_TYPE, this.getType());

        this.setValue(element, undefined);

        return element;
    }

};