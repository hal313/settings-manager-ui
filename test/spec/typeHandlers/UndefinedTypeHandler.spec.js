import { UndefinedTypeHandler } from '../../../src/typeHandlers/UndefinedTypeHandler.js';

describe('UndefinedTypeHandler', () => {

    describe('Lifecycle', () => {

        test('should exist', () => {
            expect(UndefinedTypeHandler).toBeDefined();
        });

    });

    describe('API', () => {
        let undefinedTypeHandler;

        beforeEach(() => {
            undefinedTypeHandler = new UndefinedTypeHandler();
        });

        describe('createElement', () => {

            test('should throw when name is undefined', () => {
                expect(() => undefinedTypeHandler.createElement(undefined, 'someValue')).toThrowError();
            });

            test('should throw when name is null', () => {
                expect(() => undefinedTypeHandler.createElement(null, 'someValue')).toThrowError();
            });

            test('should throw when name is the empty string', () => {
                expect(() => undefinedTypeHandler.createElement('', 'someValue')).toThrowError();
            });

            test('should throw when name is whitespace', () => {
                expect(() => undefinedTypeHandler.createElement('   ', 'someValue')).toThrowError();
            });

            test('should create an undefined value element', () => {
                const name = 'fieldName';

                // Create the element
                const element = undefinedTypeHandler.createElement(name, undefined);

                // Check the type
                expect(element instanceof Element).toBeTruthy();
                // Check the value
                expect(element.value).toEqual(undefined);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

        describe('getValue', () => {

            test('should get the value from a self-created element', () => {
                const name = 'fieldName';

                // Create the element
                const element = undefinedTypeHandler.createElement(name, undefined);

                // Check the value
                expect(undefinedTypeHandler.getValue(element)).toEqual(undefined);
            });

        });

        describe('setValue', () => {

            test('should set the value for a self-created element', () => {
                const name = 'fieldName';

                // Create the element
                const element = undefinedTypeHandler.createElement(name, undefined);

                // Precondition check
                expect(undefinedTypeHandler.getValue(element)).toEqual(undefined);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                undefinedTypeHandler.setValue(element, 'not undefined value');

                // Check the new value
                expect(undefinedTypeHandler.getValue(element)).toEqual(undefined);

                // Check the value
                expect(element.value).toEqual(undefined);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

    });

});