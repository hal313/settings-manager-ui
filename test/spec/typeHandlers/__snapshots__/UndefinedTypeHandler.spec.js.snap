// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndefinedTypeHandler API createElement should create an undefined value element 1`] = `

<div name="fieldName"
     value="undefined"
>
</div>

`;

exports[`UndefinedTypeHandler API setValue should set the value for a self-created element 1`] = `

<div name="fieldName"
     value="undefined"
>
</div>

`;

exports[`UndefinedTypeHandler API setValue should set the value for a self-created element 2`] = `

<div name="fieldName"
     value="undefined"
>
</div>

`;
